content/blog/2020-05-22-rt-template/index.md

```diff
@@ -0,0 +1,365 @@
+    Value* args[] = {i32_configs};
+    Value* ret =  builder.CreateCall(rtFunc, args);
+    
+    /* create a jump to next block */
+    Instruction *brInst = BranchInst::Create(succBB, pb);
+    
+    /* update jump from previous block */
+    Instruction *brPred = predBB->getTerminator();
+    brPred->setSuccessor(0, pb);
+```
+However, this also conflates challenges from both prior levels of abstraction.
+Unlike a single instruction, it's highly program dependent how a basic block interacts with
+rest of the program. Unlike a function, it is not clear what those program specific interfaces
+are, and has to be created at an intermediate-representation level. To make matters worse,
+these interfaces are intermediate-repesentation specific. If you target a specific optimization
+(say -O1), your pass will probably need updating for a different optimization.
+This is all great, if you're focusing on the compiler optimizations. But for a hardware 
+accelerator simulation nothing can be worse, now you are entangled in the compiler optimization
+with the simulator for performance.
+
+## Replacing chains of instructions
+One way to strike the middle ground provided by basic blocks, without the baggage from a
+particular compiler optimization, is to critically select a set of instructions to 
+replace that gives you sufficient coarseness in granularity but also decouples from the 
+intermediate representation.
+
+I motivated this by selecting chains of floating point instructions.
+
+```c
+    /* create the function call from runtime library */
+    LLVMContext& Ctx = F.getContext();
+    FunctionCallee rtFunc = F.getParent()->getOrInsertFunction(
+        "chain_replace", Type::getDoubleTy(Ctx),
+        Type::getDoubleTy(Ctx), Type::getDoubleTy(Ctx)
+    );
+    // function call name and argument types have to be known
+
+    for (auto &B: F) {
+        list <Instruction*> Selected;
+        for (auto &I: B) {
+        /* Add runtime sub block selection here */
+    
+        /* End of selection algorithm */
+    	}
+    
+        Value* args[4];
+        /* Enter instructions to the new basic block */
+        for (auto &I: Selected) {
+            IRBuilder<> builder(I);
+            builder.SetInsertPoint(&B, ++builder.GetInsertPoint());
+            Value* ret =  builder.CreateCall(rtFunc, args);
+    
+            /* forward return to users */
+            for (auto& U : I->uses()) {
+                User* user = U.getUser();
+                user->setOperand(U.getOperandNo(), ret);
+            }
+        }
+        /* Done entering new instructions */
+        
+        /* Remove replaced instructions from the basic block */
+        while (!Selected.empty()) {
+            Selected.front()->eraseFromParent();
+            Selected.pop_front();
+        }
+    }
+```
+The tradeoff with this approach is, this abstraction is an arbitrary one we have picked.
```

Comment:
Again, it's hard to tell what "this approach" is without an explanation of what the transformation is.